# A quick tour of IPython Notebook

This tour will work a little better in interactive mode, so it'll be better if you get IPython notebook installed and running. You can start it from a terminal by running

`ipython notebook`

First, we need to explain how to run cells. Try to run the cell below!

In [2]:
import pandas as pd

print("Hi! This is a cell. Press the ▶ button above to run it")

Hi! This is a cell. Press the ▶ button above to run it


You can also run a cell with Ctrl+Enter or Shift+Enter. Experiment a bit with that.

One of the most useful things about IPython notebook is its tab completion. 

Try this: click just after read_csv( in the cell below and press Shift+Tab (or Tab if you're using IPython 1.x) 4 times, slowly

This first approach assumes separator is `,`

In [7]:
pd.read_csv("biostats.csv") 

Name       "Sex"   "Age"   "Height (in)"   "Weight (lbs)"
0   Alex         "M"      41              74              170
1   Bert         "M"      42              68              166
2   Carl         "M"      32              70              155
3   Dave         "M"      39              72              167
4   Elly         "F"      30              66              124
5   Fran         "F"      33              66              115
6   Gwen         "F"      26              64              121
7   Hank         "M"      30              71              158
8   Ivan         "M"      53              72              175
9   Jake         "M"      32              69              143
10  Kate         "F"      47              69              139
11  Luke         "M"      34              72              163
12  Myra         "F"      23              62               98
13  Neil         "M"      36              75              160
14  Omar         "M"      38              70              145
15  Page         "F"      31              67              135
16  Quin         "M"      29              71              176
17  Ruth         "F"      28              65              131

This approach tell the red_csv method that the separator is `\t` i.e. `tabs`

In [8]:
pd.read_csv("biostats.csv",sep="\t") 

Name,     "Sex", "Age", "Height (in)", "Weight (lbs)"
0           Alex,       "M",   41,       74,      170   
1           Bert,       "M",   42,       68,      166   
2           Carl,       "M",   32,       70,      155   
3           Dave,       "M",   39,       72,      167   
4           Elly,       "F",   30,       66,      124   
5           Fran,       "F",   33,       66,      115   
6           Gwen,       "F",   26,       64,      121   
7           Hank,       "M",   30,       71,      158   
8           Ivan,       "M",   53,       72,      175   
9           Jake,       "M",   32,       69,      143   
10          Kate,       "F",   47,       69,      139   
11          Luke,       "M",   34,       72,      163   
12          Myra,       "F",   23,       62,       98   
13          Neil,       "M",   36,       75,      160   
14          Omar,       "M",   38,       70,      145   
15          Page,       "F",   31,       67,      135   
16          Quin,       "M",   29,       71,      176   
17          Ruth,       "F",   28,       65,      131

# Writing code

Writing code in the notebook is pretty normal.

In [10]:
def print_10_nums():
    for i in range(10):
        print(i, end=',')

In [11]:
print_10_nums()

0,1,2,3,4,5,6,7,8,9,

# Saving

As of the latest stable version, the notebook autosaves. You should use the latest stable version. Really.

# Magic functions

IPython has all kinds of magic functions. Here's an example of comparing `sum()` with a list comprehension to a generator comprehension using the `%time` magic.

In [12]:
%time sum([x for x in range(100000)])

CPU times: user 5.43 ms, sys: 4.32 ms, total: 9.75 ms
Wall time: 11.1 ms


4999950000

In [13]:
%time sum(x for x in range(100000))

CPU times: user 5.4 ms, sys: 122 µs, total: 5.52 ms
Wall time: 5.5 ms


4999950000

The magics I use most are `%time` and `%prun` for profiling. You can run `%magic` to get a list of all of them, and `%quickref` for a reference sheet.

In [14]:
%quickref

You can also do nutty things like run Perl code in the notebook with cell magics. This is especially cool for things like Cython code, where you can try out Cython really fast with the `%%cython` magic (you'll need to install it).

In [15]:
%%perl

$_ = "whoa, python!";
s/python/perl/;
print

whoa, perl!

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  